api_docs/message-formatting.md

```diff
@@ -23,6 +23,95 @@ for syntax highlighting. This field is used in the
+     </a>
+</div>
+```
+
+If the server has not yet generated thumbnails for the image yet at
+the time the message is sent, the `img` element will be a temporary
+loading indicator image and have the `image-loading-placeholder`
+class, which clients can use to identify loading indicators and
+replace them with a custom loading indicator element if desired. For
+example:
+
+``` html
+<div class="message_inline_image">
+    <a href="/user_uploads/path/to/image.png" title="image.png">
+        <img class="image-loading-placeholder" src="/path/to/spinner.png">
+     </a>
+</div>
+```
+
+Once the server has a working thumbnail, such messages will be updated
+via an `update_message` event, with the `rendering_only: true` flag
+(telling clients not to adjust message edit history), with appropriate
+adjusted `rendered_content`. A client should process those events by
+just using the updated rendering. If thumbnailing failed, the same
+type of event will edit the message's rendered form to remove the
+image preview element, so no special client-side logic should be
+required to process such errors.
+
+Note that in the uncommon situation that the thumbnailing system is
+backlogged, an individual message containing multiple image previews
+may be rerendered multiple times as each image finishes thumbnailing
+and triggers a rendering update.
+
+Clients are recommended to do the following when processing image
+previews:
+
+- If the client would like to control the thumbnail resolution used,
+  it can replace the final section of the URL (`/840x560.webp` in the
+  example above) with its preferred format from the set of supported
+  formats provided by the server in the `server_thumbnail_formats`
+  portion of the `register` response. Clients should not make any
+  assumptions about what format the server will use as the "default"
+  thumbnail resolution.
+- Download button type elements should provide the original image
+  (encoded via the `href` of the containing `a` tag).
+- Lightbox elements for viewing an image should be designed to
+  immediately display any already-downloaded thumbnail while fetching
+  the original-quality image or an appropriate higher-quality
+  thumbnail from the server, to be transparently swapped in once it is
+  available. Clients that would like to size the lightbox based on the
+  size of the original image can use the `data-original-dimensions`
+  attribute, which encodes the dimensions of the original image as
+  `{width}x{height}`, to do so.
+- Animated images will have a `data-animated` attribute on the `img`
+  tag. As detailed in `server_thumbnail_formats`, both animated and
+  still images are available for clients to use, depending on their
+  preference. See, for example, the [web
+  setting](/help/allow-image-link-previews#configure-how-animated-images-are-played)
+  to control whether animated images are autoplayed in the message
+  feed.
+- No other processing of the URLs is recommended.
```

Comment:
I think we should probably document that in rare cases, the image returned may not match the requested format (e.g. dimensions, file format, etc).  In the current protocol, this is only possible if:
- The client has a list of formats [A, B, C]
- The server updates to formats [B, C]
- A message is sent, and generates only thumbnails in [B, C]
- The client requests format A for that message

Reply:
Yeah, that sounds correct.